File: leet_code/go/partitionLabels.go
func partitionLabels(S string) []int {
	count := make(map[rune]int)
	remain, pos := len(S), 0
	var partitions []int
	for i := 0; i < len(S); i++ {
		if _, ok := count[rune(S[i])]; ok {
			count[rune(S[i])] += 1
		} else {
			count[rune(S[i])] = 1
		}
	}
	for remain > 0 {
		var part []rune
		part_count := make(map[rune]int)
		part_complete := false
		for !part_complete {
			a := rune(S[pos])
			pos += 1
			remain -= 1
			count[a] -= 1
			if _, ok := part_count[a]; ok {
				part_count[a] += 1
			} else {
				part_count[a] = 1
			}
			part_count[a] += 1
			part = append(part, a)
			if count[a] == 0 {
				finished := true
				for k, _ := range part_count {
					if count[k] != 0 {
						finished = false
					}
				}
				if finished {
					part_complete = true
				}
			}
		}
		partitions = append(partitions, len(part))
	}
	return partitions
}

File: leet_code/ruby/partition_labels.rb
# @param {String} s
# @return {Integer[]}
def partition_labels(s)
	count = Hash.new(0)
	remain = s.size
	s.each_char { |a| count[a] += 1 }
	partitions = []
	str = s.reverse.split('')
	until remain == 0
		part = []
		part_count = Hash.new(0)
		part_complete = false
		until part_complete
			a = str.pop
			remain -= 1
			count[a] -= 1
			part_count[a] += 1
			part << a
			part_complete = true if count[a] == 0 && part_count.keys.all? { |k| count[k] == 0 }
		end
		partitions << part

	end
	partitions.map { |part| part.size }
end